# <u>**Extract, Transform and Load (ETL)**</u>

This notebook is the intial step toward a standard data analytics workflow. Data quality is key to ensure effective Exploratory Data Analysis, Feature Engineering, Machine Learning Model development and dashboard visualisation. 

The data cleaning process requires thorough checks to ensure the integrity of the datasets are reviewed before saving transformations into a joined DataFrame. The steps taken in this notebook are:

- Load libraries and datasets

- Initial datasets review 

- Standardise column names 

- Ammend DataTypes 

- Identify missing, unique or duplicate values

- Add year columns for year-on-year analysis

- Join 2023-2024 and 2025-2025 datasets

- Drop identifiable information columns  

- Save DataFrame into clean data folder

ETL can reveal additional demands from the datasets during these checks. This plan is based on the most common steps taught by Code Institute LMS content, tutors and online resources included in the credits section of the README. Built-in Visual Studio Code Copilot will also collaboratively fix, modify and add comments to streamline the process.

</b>

## <u>**Load Libraries and Data**</u>

Libraries - Pandas and Numpy - are imported. Raw datasets exported as CSV from EXCEL sheets are loaded into two seperate DataFrames using Pandas.  

In [28]:
# Import libraries 
import pandas as pd
import numpy as np

# Read the CSV file into a DataFrame
df_24 = pd.read_csv('/Users/admin/Documents/Code Institute/ace_project/data/raw/2023_24_ace_raw.csv')
df_26 = pd.read_csv('/Users/admin/Documents/Code Institute/ace_project/data/raw/2025_26_ace_raw.csv')

# Check rows and columns in the DataFrames
print("Shape of df_24:", df_24.shape)
print("Shape of df_26:", df_26.shape)

Shape of df_24: (3085, 11)
Shape of df_26: (2369, 11)


</b>

### <u>**Intial DataFrame Review**</u>

Use <code>.info</code> and <code>.head()</code> for an intial review on both DataFrames.

In [29]:
df_24.info()  # Display information about df_24
df_24.head()  # Display the first few rows of df_24

<class 'pandas.DataFrame'>
RangeIndex: 3085 entries, 0 to 3084
Data columns (total 11 columns):
 #   Column                 Non-Null Count  Dtype
---  ------                 --------------  -----
 0   Recipient              3085 non-null   str  
 1   Activity name          3085 non-null   str  
 2   Award amount           3085 non-null   str  
 3   Award date             3085 non-null   str  
 4   Decision Month         3085 non-null   str  
 5   Decision quarter       3085 non-null   str  
 6   ACE Area               3085 non-null   str  
 7   Local authority        3085 non-null   str  
 8   Main discipline        3085 non-null   str  
 9   Strand                 3085 non-null   str  
 10  Time-Limited Priority  240 non-null    str  
dtypes: str(11)
memory usage: 265.2 KB


,Recipient,Activity name,Award amount,Award date,Decision Month,Decision quarter,ACE Area,Local authority,Main discipline,Strand,Time-Limited Priority
0,Feelprime Ltd,Old Fire Station Digital,"£27,689",03-Apr-23,April,Q1,South West,"Bournemouth, Christchurch and Poole",Music,"£30,000 and below",Supporting Grassroots Live Music
1,Downtown Pompey,Sticky Floors and Palm Trees,"£29,623",03-Apr-23,April,Q1,South West,Portsmouth,Theatre,"£30,000 and below",NaN
2,Chew Valley Performing Arts,Autumn Fringe Festival 2023,"£25,256",03-Apr-23,April,Q1,South West,Bath and North East Somerset,Combined arts,"£30,000 and below",NaN
3,Upfest,Weston Wallz 2023,"£30,500",03-Apr-23,April,Q1,South West,"Bristol, City of",Visual arts,"£30,000 and below",NaN
4,Chloe Thurlow,"Haunt, Pray, Love","£20,060",03-Apr-23,April,Q1,South West,South Gloucestershire,Theatre,"£30,000 and below",NaN


In [30]:
df_26.info()  # Display information about df_26
df_26.head()  # Display the first few rows of df_26

<class 'pandas.DataFrame'>
RangeIndex: 2369 entries, 0 to 2368
Data columns (total 11 columns):
 #   Column                 Non-Null Count  Dtype
---  ------                 --------------  -----
 0   Recipient              2369 non-null   str  
 1   Activity name          2369 non-null   str  
 2   Award amount           2369 non-null   str  
 3   Award date             2369 non-null   str  
 4   Decision Month         2369 non-null   str  
 5   Decision quarter       2369 non-null   str  
 6   ACE Area               2369 non-null   str  
 7   Local authority        2368 non-null   str  
 8   Main discipline        2369 non-null   str  
 9   Strand                 2369 non-null   str  
 10  Time-Limited Priority  224 non-null    str  
dtypes: str(11)
memory usage: 203.7 KB


,Recipient,Activity name,Award amount,Award date,Decision Month,Decision quarter,ACE Area,Local authority,Main discipline,Strand,Time-Limited Priority
0,Window Wanderland CIC,"Community of practise, Wanderland Pilot","£33,522",01-Apr-25,April,Q1,South West,"Bristol, City of",Combined arts,"£30,000 and below",NaN
1,Cat and Mouse Theatre,Time Crime,"£27,461",01-Apr-25,April,Q1,South West,South Gloucestershire,Theatre,"£30,000 and below",NaN
2,London Borough of Ealing Libraries and communi...,Y-Our Story Visual Stories of Acton,"£27,000",01-Apr-25,April,Q1,London,Ealing,Libraries,"£30,000 and below",Universal Library Offers
3,Exmouth Festival,Exmouth Festival 2025,"£38,500",01-Apr-25,April,Q1,South West,East Devon,Combined arts,"Above £30,000",NaN
4,Crediton Heart Project,Crediton Heart Summer Arts Festival,"£15,135",01-Apr-25,April,Q1,South West,Mid Devon,Combined arts,"£30,000 and below",NaN


The outputs for <code>.info()</code> shows both DataFrames share the same titles for column names but will require text standardisation. 

Missing values are present in both DataFrames <code>df_24</code> shows 240 rows contain values in Time-Limited Priority column. Meanwhile, <code>df_26</code> has 224 values for Time-Limited Priority and one missing value in the Local Authority column. 

DataTypes are all set as strings, which require ammending into numerical, data-time and categorical types to plot visualisation in the EDA notebook.

Identifiable information is shown in the Recipient and Activity name from the output for <code>.head()</code>. These columns must be dropped to ensure ethical data handling. 

</b>

## <u>**Standardise Column Names**</u>

Format must be consistent to access columns in the EDA, Feature Engineering and ML Modeling notebooks. Remove leading or trailing whitespace use <code>.strip()</code>, replace spaces or hyphens with underscores use <code>.replace()</code> and convert text to lowercase use <code>.lower()</code>. These functions are a professional standard protocol for technical users to navigate the code cells.  

In [31]:
# Format column names
df_24.columns = (df_24.columns
                 .str.strip()
                 .str.lower()
                 .str.replace(' ', '_'))

df_26.columns = (df_26.columns
                 .str.strip()
                 .str.lower()
                 .str.replace(' ', '_'))


In [32]:
# Check column names 2023-2024
df_24.head() 

,recipient,activity_name,award_amount,award_date,decision_month,decision_quarter,ace_area,local_authority,main_discipline,strand,time-limited_priority
0,Feelprime Ltd,Old Fire Station Digital,"£27,689",03-Apr-23,April,Q1,South West,"Bournemouth, Christchurch and Poole",Music,"£30,000 and below",Supporting Grassroots Live Music
1,Downtown Pompey,Sticky Floors and Palm Trees,"£29,623",03-Apr-23,April,Q1,South West,Portsmouth,Theatre,"£30,000 and below",NaN
2,Chew Valley Performing Arts,Autumn Fringe Festival 2023,"£25,256",03-Apr-23,April,Q1,South West,Bath and North East Somerset,Combined arts,"£30,000 and below",NaN
3,Upfest,Weston Wallz 2023,"£30,500",03-Apr-23,April,Q1,South West,"Bristol, City of",Visual arts,"£30,000 and below",NaN
4,Chloe Thurlow,"Haunt, Pray, Love","£20,060",03-Apr-23,April,Q1,South West,South Gloucestershire,Theatre,"£30,000 and below",NaN


In [33]:
# Check column names 2025-2026
df_24.head()  

,recipient,activity_name,award_amount,award_date,decision_month,decision_quarter,ace_area,local_authority,main_discipline,strand,time-limited_priority
0,Feelprime Ltd,Old Fire Station Digital,"£27,689",03-Apr-23,April,Q1,South West,"Bournemouth, Christchurch and Poole",Music,"£30,000 and below",Supporting Grassroots Live Music
1,Downtown Pompey,Sticky Floors and Palm Trees,"£29,623",03-Apr-23,April,Q1,South West,Portsmouth,Theatre,"£30,000 and below",NaN
2,Chew Valley Performing Arts,Autumn Fringe Festival 2023,"£25,256",03-Apr-23,April,Q1,South West,Bath and North East Somerset,Combined arts,"£30,000 and below",NaN
3,Upfest,Weston Wallz 2023,"£30,500",03-Apr-23,April,Q1,South West,"Bristol, City of",Visual arts,"£30,000 and below",NaN
4,Chloe Thurlow,"Haunt, Pray, Love","£20,060",03-Apr-23,April,Q1,South West,South Gloucestershire,Theatre,"£30,000 and below",NaN


</b>

## **<u>Convert DataTypes</u>**

Data consistency is neccessary to plot visualisations in the EDA notebook. Review DataTypes and convert each column to numerical, date-time or categorical. Use </code><code>.dtypes()</code> to display the current DataTypes in each DateFrame. 

The DataFrames contain mostly categorical columns, one numeric value and date-time column which were identified as strings in the intial review of both DataFrames.

In [34]:
# Check DataTypes for 2023-2024 and 2025-2026
print("Data types for df_24:")
print(df_24.dtypes)

print("\nData types for df_26:")
print(df_26.dtypes)

Data types for df_24:
recipient                str
activity_name            str
award_amount             str
award_date               str
decision_month           str
decision_quarter         str
ace_area                 str
local_authority          str
main_discipline          str
strand                   str
time-limited_priority    str
dtype: object

Data types for df_26:
recipient                str
activity_name            str
award_amount             str
award_date               str
decision_month           str
decision_quarter         str
ace_area                 str
local_authority          str
main_discipline          str
strand                   str
time-limited_priority    str
dtype: object


</b >

</b >